incremental-font-transfer/src/font_patch.rs

```diff
@@ -127,14 +156,41 @@ impl IncrementalFontPatchBase for FontRef<'_> {
             return Err(PatchingError::NonIncrementalFont);
         }
 
-        patch.apply_to(self)
+        let first = patches.first().ok_or(PatchingError::InvalidPatch(
+            "At least one patch must be provided.",
+        ))?;
+
+        match first {
+            IncrementalFontPatch::TableKeyed(patch_data) => {
+                if patches.len() > 1 {
+                    // table keyed patches can only be applied one at a time
+                    // (as they are at a minimum partial invalidating)
+                    return Err(PatchingError::IncompatiblePatches);
+                }
+                apply_table_keyed_patch(patch_data, self)
+            }
+            IncrementalFontPatch::GlyphKeyed(_) => {
+                let glyph_keyed_patches: Vec<_> = patches
+                    .iter()
+                    .filter_map(|patch| match patch {
+                        IncrementalFontPatch::GlyphKeyed(patch_data) => Some(patch_data.clone()),
+                        _ => None,
+                    })
+                    .collect();
+                // glyph keyed patches are no invalidation so any number of them may be applied simultaneously
+                if glyph_keyed_patches.len() != patches.len() {
+                    return Err(PatchingError::IncompatiblePatches);
+                }
+                apply_glyph_keyed_patch(&glyph_keyed_patches, self)
+            }
```

Comment:
see above: looking at this big whack of validation code I wonder if we're missing a chance to represent this in the type system.